In [1]:
import pandas as pd
import sqlalchemy as sa
import psycopg2 as ps
from sqlalchemy import create_engine

In [2]:
%load_ext sql
%sql postgresql://postgres:lingga28@localhost:2828/datacamp
conn = create_engine('postgresql://postgres:lingga28@localhost/datacamp')

# 1. First account for each country.
Conduct an analysis to see when the first customer accounts were created for each country.

### Instructions
- Create a table with a row for each country and columns for the country name and the date when the first customer account was created.
- Use the alias first_account for the column with the dates.
- Order by date in ascending order.

In [3]:
%%sql

SELECT country, -- For each country report the earliest date when an account was created
	MIN(date_account_start) AS first_account
FROM customers
GROUP BY country
ORDER BY first_account;

 * postgresql://postgres:***@localhost:2828/datacamp
11 rows affected.


[('France', datetime.date(2017, 1, 13)),
 ('Hungary', datetime.date(2017, 1, 18)),
 ('Belgium', datetime.date(2017, 1, 28)),
 ('Slovenia', datetime.date(2017, 1, 31)),
 ('Spain', datetime.date(2017, 2, 14)),
 ('Italy', datetime.date(2017, 2, 28)),
 ('Poland', datetime.date(2017, 3, 3)),
 ('Great Britan', datetime.date(2017, 3, 31)),
 ('Denmark', datetime.date(2017, 4, 30)),
 ('USA', datetime.date(2017, 9, 13)),
 ('Austria', datetime.date(2017, 11, 22))]

# 2. Average movie ratings
### Exercises
For each movie the average rating, the number of ratings and the number of views has to be reported. Generate a table with meaningful column names.

#### task 1
### Instruction
Group the data in the table renting by movie_id and report the ID and the average rating.

In [5]:
%%sql

SELECT movie_id, 
       AVG(rating)    -- Calculate average rating per movie
FROM renting
GROUP BY movie_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(54, Decimal('8.1666666666666667')),
 (29, Decimal('8.0000000000000000')),
 (71, Decimal('8.0000000000000000')),
 (68, Decimal('6.3333333333333333')),
 (4, Decimal('7.8888888888888889')),
 (34, Decimal('8.0000000000000000')),
 (51, Decimal('8.4285714285714286')),
 (70, Decimal('7.7500000000000000')),
 (52, Decimal('7.1428571428571429')),
 (67, Decimal('7.6666666666666667'))]

### task 2
### Instruction
- Add two columns for the number of ratings and the number of movie rentals to the results table.
- Use alias names avg_rating, number_rating and number_renting for the corresponding columns.

In [8]:
%%sql

SELECT movie_id, 
       AVG(rating) AS avg_rating, -- Use as alias avg_rating
       COUNT(rating) AS number_rating,  -- Add column for number of ratings with alias number_rating
       COUNT(date_renting) AS number_renting   -- Add column for number of movie rentals with alias number_renting
FROM renting
GROUP BY movie_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(54, Decimal('8.1666666666666667'), 6, 9),
 (29, Decimal('8.0000000000000000'), 7, 11),
 (71, Decimal('8.0000000000000000'), 4, 6),
 (68, Decimal('6.3333333333333333'), 3, 7),
 (4, Decimal('7.8888888888888889'), 9, 14),
 (34, Decimal('8.0000000000000000'), 6, 8),
 (51, Decimal('8.4285714285714286'), 7, 7),
 (70, Decimal('7.7500000000000000'), 4, 7),
 (52, Decimal('7.1428571428571429'), 7, 12),
 (67, Decimal('7.6666666666666667'), 3, 6)]

### task 3
### Instruction
- Order the rows of the table by the average rating such that it is in decreasing order.
- Observe what happens to NULL values.

In [7]:
%%sql

SELECT movie_id, 
       AVG(rating) AS avg_rating,
       COUNT(rating) AS number_ratings,
       COUNT(*) AS number_renting
FROM renting
GROUP BY movie_id
ORDER BY avg_rating DESC -- Order by average rating in decreasing order
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(11, None, 0, 6),
 (48, Decimal('10.0000000000000000'), 2, 5),
 (5, Decimal('9.7500000000000000'), 4, 8),
 (42, Decimal('9.6000000000000000'), 5, 10),
 (63, Decimal('9.3333333333333333'), 6, 11),
 (3, Decimal('9.0000000000000000'), 2, 11),
 (65, Decimal('9.0000000000000000'), 5, 5),
 (18, Decimal('9.0000000000000000'), 5, 7),
 (57, Decimal('8.8333333333333333'), 6, 9),
 (6, Decimal('8.8000000000000000'), 5, 6)]

### task 4
### Instruction
Question
Which statement is true for the movie with average rating null?

### Possible Answers:
- A. The number of ratings is 6.
- B. The number of movie rentals is zero.
- C. The average is null because one of the ratings of the movie is null.
- D. The average is null because all of the ratings of the movie are null.

Answer: D.

# 3. Average rating per customer
### Exercises
Similar to what you just did, you will now look at the average movie ratings, this time for customers. So you will obtain a table with the average rating given by each customer. Further, you will include the number of ratings and the number of movie rentals per customer. You will report these summary statistics only for customers with more than 7 movie rentals and order them in ascending order by the average rating.

### Instructions
- Group the data in the table renting by customer_id and report the customer_id, the average rating, the number of ratings and the number of movie rentals.
- Select only customers with more than 7 movie rentals.
- rder the resulting table by the average rating in ascending order.

In [9]:
%%sql

SELECT customer_id, -- Report the customer_id
      AVG(rating),  -- Report the average rating per customer
      COUNT(rating),  -- Report the number of ratings per customer
      COUNT(*)  -- Report the number of movie rentals per customer
FROM renting
GROUP BY customer_id
HAVING COUNT(*) > 7 -- Select only customers with more than 7 movie rentals
ORDER BY AVG(rating); -- Order by the average rating in ascending order

 * postgresql://postgres:***@localhost:2828/datacamp
22 rows affected.


[(104, Decimal('6.2500000000000000'), 4, 8),
 (28, Decimal('6.7142857142857143'), 7, 11),
 (111, Decimal('7.0000000000000000'), 3, 10),
 (113, Decimal('7.0000000000000000'), 7, 15),
 (25, Decimal('7.2000000000000000'), 5, 10),
 (21, Decimal('7.3333333333333333'), 6, 14),
 (92, Decimal('7.5714285714285714'), 7, 11),
 (49, Decimal('7.6250000000000000'), 8, 13),
 (35, Decimal('7.6666666666666667'), 6, 9),
 (52, Decimal('7.8750000000000000'), 8, 9),
 (108, Decimal('8.0000000000000000'), 6, 10),
 (114, Decimal('8.0000000000000000'), 6, 11),
 (66, Decimal('8.0000000000000000'), 7, 10),
 (93, Decimal('8.4000000000000000'), 5, 8),
 (26, Decimal('8.4000000000000000'), 5, 10),
 (94, Decimal('8.7142857142857143'), 7, 10),
 (121, Decimal('8.7500000000000000'), 4, 9),
 (7, Decimal('9.0000000000000000'), 4, 9),
 (61, Decimal('9.2000000000000000'), 5, 8),
 (8, Decimal('9.2500000000000000'), 4, 8),
 (99, Decimal('9.2500000000000000'), 4, 8),
 (65, Decimal('9.3333333333333333'), 3, 9)]

# 4. Join renting and customers
### Exercises
For many analyses it is necessary to add customer information to the data in the table renting.

### task 1
### Instruction
- Augment the table renting with all columns from the table customers with a LEFT JOIN.
- Use as alias' for the tables r and c respectively.

In [10]:
%%sql

SELECT * -- Join renting with customers
FROM renting AS r
LEFT JOIN customers AS c
ON c.customer_id = r.customer_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(1, 41, 8, None, datetime.date(2018, 10, 9), 41, 'Zara Mitchell', 'Great Britan', 'female', datetime.date(1994, 7, 8), datetime.date(2017, 6, 12)),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28)),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 108, 'Saúl Tafoya Meraz', 'Spain', 'male', datetime.date(1992, 5, 15), datetime.date(2017, 3, 13)),
 (4, 39, 66, 8, datetime.date(2018, 10, 22), 39, 'Amy Haynes', 'Great Britan', 'female', datetime.date(1975, 7, 28), datetime.date(2018, 1, 19)),
 (5, 104, 15, 7, datetime.date(2019, 3, 18), 104, 'Dimitri Zambrano Morales', 'Spain', 'male', datetime.date(1992, 6, 21), datetime.date(2017, 4, 12)),
 (6, 50, 71, 7, datetime.date(2018, 10, 9), 50, 'Csordás Ármin', 'Hungary', 'male', datetime.date(1986, 11, 30), datetime.date(2018, 4, 23)),
 (7, 52, 21, None, datetime.date(2018, 11, 10), 52, 'Szepessi Marton', 'Hungary', 'male', datetime.date(1983, 3, 14), datetime.date(2018, 3, 24)),
 (8, 73, 65, 10, datetime.date(2018, 6, 5), 73, 'Veneranda Milani', 'Italy', 'female', datetime.date(1983, 1, 15), datetime.date(2018, 1, 15)),
 (9, 78, 2, None, datetime.date(2017, 9, 3), 78, 'Doroteo Napolitano', 'Italy', 'male', datetime.date(1992, 6, 1), datetime.date(2017, 7, 27)),
 (10, 121, 43, None, datetime.date(2017, 11, 8), 121, 'Annie R. Barnett', 'USA', 'female', datetime.date(1987, 3, 2), datetime.date(2017, 9, 13))]

### task 2
### Instruction
Select only records from customers coming from Belgium.

In [11]:
%%sql

SELECT *
FROM renting AS r
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
WHERE country = 'Belgium'; -- Select only records from customers coming from Belgium

 * postgresql://postgres:***@localhost:2828/datacamp
24 rows affected.


[(2, 10, 29, 10, datetime.date(2017, 3, 1), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28)),
 (14, 8, 29, None, datetime.date(2018, 8, 3), 8, 'Jaëla van den Dolder', 'Belgium', 'female', datetime.date(1990, 8, 31), datetime.date(2018, 2, 8)),
 (27, 7, 36, None, datetime.date(2019, 3, 14), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12)),
 (32, 8, 42, 10, datetime.date(2019, 2, 13), 8, 'Jaëla van den Dolder', 'Belgium', 'female', datetime.date(1990, 8, 31), datetime.date(2018, 2, 8)),
 (70, 10, 44, None, datetime.date(2017, 12, 17), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28)),
 (187, 7, 44, 8, datetime.date(2018, 6, 23), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12)),
 (206, 7, 16, None, datetime.date(2019, 4, 27), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12)),
 (237, 10, 38, None, datetime.date(2018, 6, 29), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28)),
 (243, 7, 5, 10, datetime.date(2019, 1, 11), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12)),
 (252, 8, 61, None, datetime.date(2018, 8, 7), 8, 'Jaëla van den Dolder', 'Belgium', 'female', datetime.date(1990, 8, 31), datetime.date(2018, 2, 8)),
 (273, 8, 5, 10, datetime.date(2019, 1, 21), 8, 'Jaëla van den Dolder', 'Belgium', 'female', datetime.date(1990, 8, 31), datetime.date(2018, 2, 8)),
 (309, 10, 38, None, datetime.date(2018, 6, 24), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28)),
 (313, 7, 4, None, datetime.date(2018, 9, 22), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12)),
 (339, 8, 2, None, datetime.date(2018, 11, 18), 8, 'Jaëla van den Dolder', 'Belgium', 'female', datetime.date(1990, 8, 31), datetime.date(2018, 2, 8)),
 (363, 8, 33, 7, datetime.date(2018, 9, 19), 8, 'Jaëla van den Dolder', 'Belgium', 'female', datetime.date(1990, 8, 31), datetime.date(2018, 2, 8)),
 (386, 8, 16, 10, datetime.date(2018, 6, 29), 8, 'Jaëla van den Dolder', 'Belgium', 'female', datetime.date(1990, 8, 31), datetime.date(2018, 2, 8)),
 (396, 7, 40, 10, datetime.date(2018, 9, 11), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12)),
 (398, 8, 37, None, datetime.date(2018, 3, 23), 8, 'Jaëla van den Dolder', 'Belgium', 'female', datetime.date(1990, 8, 31), datetime.date(2018, 2, 8)),
 (400, 6, 10, None, datetime.date(2018, 11, 3), 6, 'Rowanne Couperus', 'Belgium', 'female', datetime.date(1994, 4, 5), datetime.date(2018, 8, 26)),
 (456, 7, 3, None, datetime.date(2018, 5, 23), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12)),
 (536, 7, 50, 8, datetime.date(2018, 9, 16), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12)),
 (555, 10, 5, None, datetime.date(2018, 12, 2), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28)),
 (560, 11, 49, 6, datetime.date(2019, 4, 18), 11, 'Yakup Mes', 'Belgium', 'male', datetime.date(1978, 4, 28), datetime.date(2019, 2, 19)),
 (574, 7, 14, None, datetime.date(2019, 2, 8), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12))]

### task 3
### Instruction
Average ratings of customers from Belgium.

In [12]:
%%sql

SELECT AVG(r.rating) -- Average ratings of customers from Belgium
FROM renting AS r
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
WHERE c.country='Belgium';

 * postgresql://postgres:***@localhost:2828/datacamp
1 rows affected.


[(Decimal('8.9000000000000000'),)]

# 5. Aggregating revenue, rentals and active customers
### Exercises
The management of MovieNow wants to report key performance indicators (KPIs) for the performance of the company in 2018. They are interested in measuring the financial successes as well as user engagement. Important KPIs are, therefore, the profit coming from movie rentals, the number of movie rentals and the number of active customers.

### task 1
### Instruction
- First, you need to join movies on renting to include the renting_price from the movies table for each renting record.
- Use as alias' for the tables m and r respectively.

In [13]:
%%sql

SELECT *
FROM renting AS r
LEFT JOIN movies AS m -- Choose the correct join statment
ON r.movie_id = m.movie_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(1, 41, 8, None, datetime.date(2018, 10, 9), 8, 'Waking Up in Reno', 'Comedy', 91, 2002, Decimal('2.59')),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 29, 'Two for the Money', 'Drama', 122, 2005, Decimal('2.79')),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 45, 'Burn After Reading', 'Drama', 96, 2008, Decimal('2.39')),
 (4, 39, 66, 8, datetime.date(2018, 10, 22), 66, 'The Hunger Games', 'Drama', 142, 2012, Decimal('1.59')),
 (5, 104, 15, 7, datetime.date(2019, 3, 18), 15, "Something's Gotta Give", 'Comedy', 128, 2003, Decimal('1.69')),
 (6, 50, 71, 7, datetime.date(2018, 10, 9), 71, 'Rio 2', 'Animation', 101, 2014, Decimal('1.79')),
 (7, 52, 21, None, datetime.date(2018, 11, 10), 21, 'The Return of the King', 'Science Fiction & Fantasy', 200, 2003, Decimal('1.99')),
 (8, 73, 65, 10, datetime.date(2018, 6, 5), 65, 'Ghost Rider: Spirit of Vengeance', 'Action & Adventure', 96, 2012, Decimal('1.79')),
 (9, 78, 2, None, datetime.date(2017, 9, 3), 2, 'Swordfish', 'Drama', 99, 2001, Decimal('2.19')),
 (10, 121, 43, None, datetime.date(2017, 11, 8), 43, 'The Invasion', 'Mystery & Suspense', 99, 2007, Decimal('2.09'))]

### task 2
### Instruction
Calculate the revenue coming from movie rentals, the number of movie rentals and the number of customers who rented a movie.

In [14]:
%%sql

SELECT 
	SUM(m.renting_price), -- Get the revenue from movie rentals
	COUNT(*), -- Count the number of rentals
	COUNT(DISTINCT r.customer_id)  -- Count the number of customers
FROM renting AS r
LEFT JOIN movies AS m
ON r.movie_id = m.movie_id;

 * postgresql://postgres:***@localhost:2828/datacamp
1 rows affected.


[(Decimal('1275.72'), 578, 116)]

### task 3
### Instruction
Now, you can report these values for the year 2018. Calculate the revenue in 2018, the number of movie rentals and the number of active customers in 2018. An active customer is a customer who rented at least one movie in 2018.

In [15]:
%%sql

SELECT 
	SUM(m.renting_price), 
	COUNT(*), 
	COUNT(DISTINCT r.customer_id)
FROM renting AS r
LEFT JOIN movies AS m
ON r.movie_id = m.movie_id
-- Only look at movie rentals in 2018
WHERE date_renting BETWEEN '2018-01-01' AND '2018-12-31';

 * postgresql://postgres:***@localhost:2828/datacamp
1 rows affected.


[(Decimal('658.02'), 298, 93)]

# 6. Movies and actors
### Exercises
You are asked to give an overview of which actors play in which movie.

### Instructions
- Create a list of actor names and movie titles in which they act. Make sure that each combination of actor and movie appears only once.
- Use as an alias for the table actsin the two letters ai.

In [16]:
%%sql

SELECT m.title, -- Create a list of movie titles and actor names
       a.name
FROM actsin AS ai
LEFT JOIN movies AS m
ON m.movie_id = ai.movie_id
LEFT JOIN actors AS a
ON a.actor_id = ai.actor_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Candy', 'Abbie Cornish'),
 ('Jack and Jill', 'Adam Sandler'),
 ('Simone', 'Al Pacino'),
 ('The Recruit', 'Al Pacino'),
 ('Two for the Money', 'Al Pacino'),
 ('Jack and Jill', 'Al Pacino'),
 ('The Fighter', 'Amy Adams'),
 ('Brighton Rock', 'Andrea Riseborough'),
 ('The Royal Tenenbaums', 'Anjelica Huston'),
 ('Sleepwalking', 'Annasophia Robb')]

# 7. Income from movies
### Exercises
How much income did each movie generate? To answer this question subsequent SELECT statements can be used.

### task 1
### Instruction
- Use a join to get the movie title and price for each movie rental.

In [17]:
%%sql

SELECT m.title, -- Use a join to get the movie title and price for each movie rental
       m.renting_price
FROM renting AS r
LEFT JOIN movies AS m
ON r.movie_id = m.movie_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Waking Up in Reno', Decimal('2.59')),
 ('Two for the Money', Decimal('2.79')),
 ('Burn After Reading', Decimal('2.39')),
 ('The Hunger Games', Decimal('1.59')),
 ("Something's Gotta Give", Decimal('1.69')),
 ('Rio 2', Decimal('1.79')),
 ('The Return of the King', Decimal('1.99')),
 ('Ghost Rider: Spirit of Vengeance', Decimal('1.79')),
 ('Swordfish', Decimal('2.19')),
 ('The Invasion', Decimal('2.09'))]

### task 2
### Instruction
- Report the total income for each movie.
- Order the result by decreasing income.

In [18]:
%%sql

SELECT rm.title, -- Report the income from movie rentals for each movie 
       SUM(rm.renting_price) AS income_movie
FROM
       (SELECT m.title, 
               m.renting_price
       FROM renting AS r
       LEFT JOIN movies AS m
       ON r.movie_id=m.movie_id) AS rm
GROUP BY rm.title
ORDER BY income_movie DESC -- Order the result by decreasing income
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Bridget Jones - The Edge of Reason', Decimal('37.57')),
 ('Fair Game', Decimal('34.68')),
 ('The Kingdom', Decimal('31.35')),
 ('Two for the Money', Decimal('30.69')),
 ('Simone', Decimal('29.59')),
 ('Django Unchained', Decimal('29.59')),
 ('What Women Want', Decimal('28.49')),
 ('Monster', Decimal('27.17')),
 ('The Two Towers', Decimal('26.29')),
 ('A Good Woman', Decimal('26.01'))]

### task 3
### Instruction
Question
Which statement about the movie 'Django Unchained' is NOT correct?

### Possible Answers
- A. The income from this movie is 29.59.
- B. The income from 'Django Unchained' is lower than from 'The Kingdom'.
- C. The income from 'Django Unchained' is higher than from 'Simone'.
- D. 'Django Unchained' has the 5th highest income.

Answer: C

# 8. Age of actors from the USA
### Exercises
Now you will explore the age of American actors and actresses. Report the date of birth of the oldest and youngest US actor and actress.

### Instructions
- Create a subsequent SELECT statements in the FROM clause to get all information about actors from the USA.
- Give the subsequent SELECT statement the alias a.
- Report for actors from the USA the year of birth of the oldest and the year of birth of the youngest actor and actress.

In [19]:
%%sql

SELECT a.gender, -- Report for male and female actors from the USA 
       MAX(a.year_of_birth), -- The year of birth of the oldest actor
       MIN(a.year_of_birth) -- The year of birth of the youngest actor
FROM
   (SELECT * -- Use a subsequen SELECT to get all information about actors from the USA
   FROM actors
   WHERE nationality = 'USA') AS a -- Give the table the name a
GROUP BY a.gender;

 * postgresql://postgres:***@localhost:2828/datacamp
2 rows affected.


[('female', 1993, 1945), ('male', 1992, 1930)]

# 9. Identify favorite movies for a group of customers
Which is the favorite movie on MovieNow? Answer this question for a specific group of customers: for all customers born in the 70s.

### task 1
### Instruction
- Augment the table renting with customer information and information about the movies.
- For each join use the first letter of the table name as alias.

In [20]:
%%sql

SELECT *
FROM renting AS r
LEFT JOIN customers AS c  -- Add customer information
ON r.customer_id = c.customer_id
LEFT JOIN movies AS m   -- Add movie information
ON r.movie_id = m.movie_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(1, 41, 8, None, datetime.date(2018, 10, 9), 41, 'Zara Mitchell', 'Great Britan', 'female', datetime.date(1994, 7, 8), datetime.date(2017, 6, 12), 8, 'Waking Up in Reno', 'Comedy', 91, 2002, Decimal('2.59')),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28), 29, 'Two for the Money', 'Drama', 122, 2005, Decimal('2.79')),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 108, 'Saúl Tafoya Meraz', 'Spain', 'male', datetime.date(1992, 5, 15), datetime.date(2017, 3, 13), 45, 'Burn After Reading', 'Drama', 96, 2008, Decimal('2.39')),
 (4, 39, 66, 8, datetime.date(2018, 10, 22), 39, 'Amy Haynes', 'Great Britan', 'female', datetime.date(1975, 7, 28), datetime.date(2018, 1, 19), 66, 'The Hunger Games', 'Drama', 142, 2012, Decimal('1.59')),
 (5, 104, 15, 7, datetime.date(2019, 3, 18), 104, 'Dimitri Zambrano Morales', 'Spain', 'male', datetime.date(1992, 6, 21), datetime.date(2017, 4, 12), 15, "Something's Gotta Give", 'Comedy', 128, 2003, Decimal('1.69')),
 (6, 50, 71, 7, datetime.date(2018, 10, 9), 50, 'Csordás Ármin', 'Hungary', 'male', datetime.date(1986, 11, 30), datetime.date(2018, 4, 23), 71, 'Rio 2', 'Animation', 101, 2014, Decimal('1.79')),
 (7, 52, 21, None, datetime.date(2018, 11, 10), 52, 'Szepessi Marton', 'Hungary', 'male', datetime.date(1983, 3, 14), datetime.date(2018, 3, 24), 21, 'The Return of the King', 'Science Fiction & Fantasy', 200, 2003, Decimal('1.99')),
 (8, 73, 65, 10, datetime.date(2018, 6, 5), 73, 'Veneranda Milani', 'Italy', 'female', datetime.date(1983, 1, 15), datetime.date(2018, 1, 15), 65, 'Ghost Rider: Spirit of Vengeance', 'Action & Adventure', 96, 2012, Decimal('1.79')),
 (9, 78, 2, None, datetime.date(2017, 9, 3), 78, 'Doroteo Napolitano', 'Italy', 'male', datetime.date(1992, 6, 1), datetime.date(2017, 7, 27), 2, 'Swordfish', 'Drama', 99, 2001, Decimal('2.19')),
 (10, 121, 43, None, datetime.date(2017, 11, 8), 121, 'Annie R. Barnett', 'USA', 'female', datetime.date(1987, 3, 2), datetime.date(2017, 9, 13), 43, 'The Invasion', 'Mystery & Suspense', 99, 2007, Decimal('2.09'))]

### task 2
### Instruction
Select only those records of customers born in the 70s.

In [21]:
%%sql

SELECT *
FROM renting AS r
LEFT JOIN customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
WHERE c.date_of_birth BETWEEN '1970-01-01' AND '1979-12-31' -- Select customers born in the 70s
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(4, 39, 66, 8, datetime.date(2018, 10, 22), 39, 'Amy Haynes', 'Great Britan', 'female', datetime.date(1975, 7, 28), datetime.date(2018, 1, 19), 66, 'The Hunger Games', 'Drama', 142, 2012, Decimal('1.59')),
 (11, 61, 61, None, datetime.date(2017, 6, 4), 61, 'Nella Manfrin', 'Italy', 'female', datetime.date(1974, 1, 11), datetime.date(2017, 5, 14), 61, 'Harry Potter and the Deathly Hallows – Part 2', 'Science Fiction & Fantasy', 130, 2011, Decimal('1.99')),
 (15, 13, 5, 10, datetime.date(2017, 11, 25), 13, 'Thomas T. Nissen', 'Denmark', 'male', datetime.date(1975, 9, 25), datetime.date(2017, 4, 30), 5, 'The Fellowship of the Ring', 'Science Fiction & Fantasy', 178, 2001, Decimal('2.59')),
 (17, 22, 46, 10, datetime.date(2019, 2, 16), 22, 'Delit Perillard', 'France', 'female', datetime.date(1970, 1, 10), datetime.date(2018, 11, 18), 46, 'Sleepwalking', 'Drama', 101, 2008, Decimal('1.79')),
 (18, 36, 39, 10, datetime.date(2019, 3, 20), 36, 'Jacob Hope', 'Great Britan', 'male', datetime.date(1979, 8, 27), datetime.date(2018, 6, 24), 39, "I'm Not There", 'Drama', 135, 2007, Decimal('1.79')),
 (21, 30, 60, 10, datetime.date(2018, 2, 20), 30, 'Gradasso Mousseau', 'France', 'male', datetime.date(1976, 7, 11), datetime.date(2017, 10, 10), 60, 'Black Death', 'Drama', 102, 2011, Decimal('2.69')),
 (22, 19, 35, None, datetime.date(2017, 5, 3), 19, 'Avelaine Piedalue', 'France', 'female', datetime.date(1972, 4, 14), datetime.date(2017, 1, 13), 35, 'Happy Feet', 'Animation', 108, 2006, Decimal('1.79')),
 (24, 57, 41, 7, datetime.date(2018, 8, 16), 57, 'Mario Lettiere', 'Italy', 'male', datetime.date(1975, 5, 4), datetime.date(2018, 1, 30), 41, 'The Kingdom', 'Drama', 110, 2007, Decimal('2.09')),
 (26, 118, 27, 8, datetime.date(2018, 12, 1), 118, 'Digna Mateo Carrasquill', 'Spain', 'female', datetime.date(1978, 11, 1), datetime.date(2017, 11, 24), 27, 'Monster', 'Drama', 109, 2004, Decimal('2.09')),
 (28, 113, 40, 7, datetime.date(2018, 4, 5), 113, 'Lucy Centeno Barrios', 'Spain', 'female', datetime.date(1970, 11, 3), datetime.date(2017, 6, 13), 40, 'Then She Found Me', 'Drama', 100, 2007, Decimal('1.99'))]

### task 3
### Instruction
For each movie, report the number of times it was rented, as well as the average rating. Limit your results to customers born in the 1970s.

In [22]:
%%sql

SELECT m.title, 
COUNT(*), -- Report number of views per movie
AVG(r.rating) -- Add the average rating for each movie
FROM renting AS r
LEFT JOIN customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
WHERE c.date_of_birth BETWEEN '1970-01-01' AND '1979-12-31'
GROUP BY m.title
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('V for Vendetta', 2, Decimal('7.0000000000000000')),
 ('The Fighter', 4, Decimal('10.0000000000000000')),
 ("Winter's Bone", 2, Decimal('9.5000000000000000')),
 ('A Good Woman', 4, Decimal('7.3333333333333333')),
 ('The Return of the King', 1, None),
 ('Secondhand Lions', 4, Decimal('9.6666666666666667')),
 ('The Frozen Ground', 4, Decimal('7.5000000000000000')),
 ('Ray', 2, None),
 ('Love Actually', 1, Decimal('7.0000000000000000')),
 ('The Company You Keep', 1, None)]

### task 4
### Instruction
- Remove those movies from the table with only one rental.
- Order the result table such that movies with highest rating come first.

In [23]:
%%sql

SELECT m.title, 
COUNT(*),
AVG(r.rating) 
FROM renting AS r
LEFT JOIN customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
WHERE c.date_of_birth BETWEEN '1970-01-01' AND '1979-12-31'
GROUP BY m.title
HAVING COUNT(*) > 1  -- Remove movies with only one rental
ORDER BY AVG(r.rating) DESC -- Order with highest rating first
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Ray', 2, None),
 ('Harry Potter and the Deathly Hallows – Part 2', 2, None),
 ('Waking Up in Reno', 2, None),
 ('Showtime', 5, None),
 ('The Fellowship of the Ring', 2, Decimal('10.0000000000000000')),
 ('No Country for Old Men', 3, Decimal('10.0000000000000000')),
 ("I'm Not There", 2, Decimal('10.0000000000000000')),
 ('The Fighter', 4, Decimal('10.0000000000000000')),
 ("One Night at McCool's", 2, Decimal('10.0000000000000000')),
 ('Django Unchained', 4, Decimal('10.0000000000000000'))]

# 10. Identify favorite actors for Spain
### Exercises
You're now going to explore actor popularity in Spain. Use as alias the first letter of the table, except for the table actsin use ai instead.

### task 1
### Instruction
Augment the table renting with information about customers and actors.

In [24]:
%%sql

SELECT *
FROM renting as r 
LEFT JOIN customers AS c  -- Augment table renting with information about customers 
ON r.customer_id = c.customer_id
LEFT JOIN actsin AS ai  -- Augment the table renting with the table actsin
ON r.movie_id = ai.movie_id
LEFT JOIN actors AS a -- Augment table renting with information about actors
ON ai.actor_id = a.actor_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(1, 41, 8, None, datetime.date(2018, 10, 9), 41, 'Zara Mitchell', 'Great Britan', 'female', datetime.date(1994, 7, 8), datetime.date(2017, 6, 12), 160, 8, 107, 107, 'Patrick Swayze', 1952, 'USA', 'male'),
 (1, 41, 8, None, datetime.date(2018, 10, 9), 41, 'Zara Mitchell', 'Great Britan', 'female', datetime.date(1994, 7, 8), datetime.date(2017, 6, 12), 152, 8, 103, 103, 'Natasha Richardson', 1963, 'British', 'female'),
 (1, 41, 8, None, datetime.date(2018, 10, 9), 41, 'Zara Mitchell', 'Great Britan', 'female', datetime.date(1994, 7, 8), datetime.date(2017, 6, 12), 28, 8, 23, 23, 'Charlize Theron', 1975, 'South Africa', 'female'),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28), 172, 29, 117, 117, 'Rene Russo', 1954, 'USA', 'female'),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28), 143, 29, 96, 96, 'Matthew McConaughey', 1969, 'USA', 'male'),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28), 5, 29, 3, 3, 'Al Pacino', 1940, 'USA', 'male'),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 108, 'Saúl Tafoya Meraz', 'Spain', 'male', datetime.date(1992, 5, 15), datetime.date(2017, 3, 13), 118, 45, 74, 74, 'John Malkovich', 1953, 'USA', 'male'),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 108, 'Saúl Tafoya Meraz', 'Spain', 'male', datetime.date(1992, 5, 15), datetime.date(2017, 3, 13), 80, 45, 49, 49, 'George Clooney', 1961, 'USA', 'male'),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 108, 'Saúl Tafoya Meraz', 'Spain', 'male', datetime.date(1992, 5, 15), datetime.date(2017, 3, 13), 74, 45, 45, 45, 'Frances McDormand', 1957, 'USA', 'female'),
 (4, 39, 66, 8, datetime.date(2018, 10, 22), 39, 'Amy Haynes', 'Great Britan', 'female', datetime.date(1975, 7, 28), datetime.date(2018, 1, 19), 130, 66, 86, 86, 'Liam Hemsworth', 1990, 'Australia', 'male')]

### task 2
### Instruction
- Report the number of movie rentals and the average rating for each actor, separately for male and female customers.
- Report only actors with more than 5 movie rentals.

In [25]:
%%sql

SELECT a.name,  c.gender,
       COUNT(*) AS number_views, 
       AVG(r.rating) AS avg_rating
FROM renting as r
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
LEFT JOIN actsin as ai
ON r.movie_id = ai.movie_id
LEFT JOIN actors as a
ON ai.actor_id = a.actor_id

GROUP BY a.name, c.gender -- For each actor, separately for male and female customers
HAVING AVG(r.rating) IS NOT NULL 
  AND COUNT(*) > 5 -- Report only actors with more than 5 movie rentals
ORDER BY avg_rating DESC, number_views DESC
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Tommy Lee Jones', 'female', 6, Decimal('10.0000000000000000')),
 ('Josh Brolin', 'female', 6, Decimal('10.0000000000000000')),
 ('Javier Bardem', 'female', 6, Decimal('10.0000000000000000')),
 ('Christoph Waltz', 'female', 8, Decimal('9.3333333333333333')),
 ('Leonardo DiCaprio', 'female', 8, Decimal('9.3333333333333333')),
 ('Barry Pepper', 'female', 7, Decimal('9.3333333333333333')),
 ('Edward Norton', 'female', 7, Decimal('9.3333333333333333')),
 ('Richard Gere', 'male', 6, Decimal('9.2000000000000000')),
 ('Rene Russo', 'female', 8, Decimal('9.0000000000000000')),
 ('Hector Bordoni', 'female', 7, Decimal('9.0000000000000000'))]

### task 3
### Instruction
Now, report the favorite actors only for customers from Spain.

In [26]:
%%sql

SELECT a.name,  c.gender,
       COUNT(*) AS number_views, 
       AVG(r.rating) AS avg_rating
FROM renting as r
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
LEFT JOIN actsin as ai
ON r.movie_id = ai.movie_id
LEFT JOIN actors as a
ON ai.actor_id = a.actor_id
WHERE c.country = 'Spain' -- Select only customers from Spain
GROUP BY a.name, c.gender
HAVING AVG(r.rating) IS NOT NULL 
  AND COUNT(*) > 5 
ORDER BY avg_rating DESC, number_views DESC;

 * postgresql://postgres:***@localhost:2828/datacamp
4 rows affected.


[('Catherine Keener', 'female', 6, Decimal('8.0000000000000000')),
 ('Emma Watson', 'male', 7, Decimal('7.6000000000000000')),
 ('Daniel Radcliffe', 'male', 7, Decimal('7.6000000000000000')),
 ('Rupert Grint', 'male', 7, Decimal('7.6000000000000000'))]

# 11. KPIs per country
### Exercises
In chapter 1 you were asked to provide a report about the development of the company. This time you have to prepare a similar report with KPIs for each country separately. Your manager is interested in the total number of movie rentals, the average rating of all movies and the total revenue for each country since the beginning of 2019.

### task 1
### Instructions
- Augment the table renting with information about customers and movies.
- Use as alias the first latter of the table name.
- Select only records about rentals since beginning of 2019.

In [27]:
%%sql

SELECT *
FROM renting AS r -- Augment the table renting with information about customers
LEFT JOIN customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies AS m -- Augment the table renting with information about movies
ON r.movie_id = m.movie_id
WHERE date_renting >= '2019-01-01' -- Select only records about rentals since the beginning of 2019
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(5, 104, 15, 7, datetime.date(2019, 3, 18), 104, 'Dimitri Zambrano Morales', 'Spain', 'male', datetime.date(1992, 6, 21), datetime.date(2017, 4, 12), 15, "Something's Gotta Give", 'Comedy', 128, 2003, Decimal('1.69')),
 (17, 22, 46, 10, datetime.date(2019, 2, 16), 22, 'Delit Perillard', 'France', 'female', datetime.date(1970, 1, 10), datetime.date(2018, 11, 18), 46, 'Sleepwalking', 'Drama', 101, 2008, Decimal('1.79')),
 (18, 36, 39, 10, datetime.date(2019, 3, 20), 36, 'Jacob Hope', 'Great Britan', 'male', datetime.date(1979, 8, 27), datetime.date(2018, 6, 24), 39, "I'm Not There", 'Drama', 135, 2007, Decimal('1.79')),
 (27, 7, 36, None, datetime.date(2019, 3, 14), 7, 'Annelous Sneep', 'Belgium', 'female', datetime.date(1993, 11, 14), datetime.date(2018, 5, 12), 36, 'World Trade Center', 'Drama', 129, 2006, Decimal('1.59')),
 (32, 8, 42, 10, datetime.date(2019, 2, 13), 8, 'Jaëla van den Dolder', 'Belgium', 'female', datetime.date(1990, 8, 31), datetime.date(2018, 2, 8), 42, 'No Country for Old Men', 'Drama', 122, 2007, Decimal('1.49')),
 (35, 91, 7, 8, datetime.date(2019, 3, 27), 91, 'Ewa Zawadzka', 'Poland', 'female', datetime.date(1988, 7, 13), datetime.date(2018, 1, 5), 7, 'The Royal Tenenbaums', 'Comedy', 110, 2002, Decimal('1.89')),
 (40, 39, 27, 8, datetime.date(2019, 3, 22), 39, 'Amy Haynes', 'Great Britan', 'female', datetime.date(1975, 7, 28), datetime.date(2018, 1, 19), 27, 'Monster', 'Drama', 109, 2004, Decimal('2.09')),
 (42, 31, 43, 6, datetime.date(2019, 2, 26), 31, 'Christien DAubigné,France,male,1986-11-06,2018-02-13\n32,Hayden ODonnell', 'Great Britan', 'male', datetime.date(1991, 11, 7), datetime.date(2017, 11, 20), 43, 'The Invasion', 'Mystery & Suspense', 99, 2007, Decimal('2.09')),
 (43, 109, 35, None, datetime.date(2019, 1, 22), 109, 'Zohar Malave Ornelas', 'Spain', 'male', datetime.date(1979, 10, 22), datetime.date(2018, 7, 22), 35, 'Happy Feet', 'Animation', 108, 2006, Decimal('1.79')),
 (44, 62, 52, 6, datetime.date(2019, 1, 4), 62, 'Gina Lucciano', 'Italy', 'female', datetime.date(1985, 11, 29), datetime.date(2018, 5, 22), 52, 'Fair Game', 'Drama', 108, 2010, Decimal('2.89'))]

### task 2
### Instruction
- Calculate the number of movie rentals.
- Calculate the average rating.
- Calculate the revenue from movie rentals.
- Report these KPIs for each country.

In [28]:
%%sql

SELECT 
	c.country,                      -- For each country report
	COUNT(*) AS number_renting,   -- The number of movie rentals
	AVG(r.rating) AS average_rating,  -- The average rating
	SUM(m.renting_price) AS revenue  -- The revenue from movie rentals
FROM renting AS r
LEFT JOIN customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
WHERE date_renting >= '2019-01-01'
GROUP BY c.country;

 * postgresql://postgres:***@localhost:2828/datacamp
12 rows affected.


[(None, 2, Decimal('8.5000000000000000'), Decimal('3.48')),
 ('Spain', 26, Decimal('8.0769230769230769'), Decimal('57.94')),
 ('Great Britan', 10, Decimal('7.3333333333333333'), Decimal('20.70')),
 ('Austria', 4, Decimal('6.0000000000000000'), Decimal('8.16')),
 ('Italy', 30, Decimal('8.2352941176470588'), Decimal('69.90')),
 ('Poland', 21, Decimal('8.1250000000000000'), Decimal('49.29')),
 ('Slovenia', 14, Decimal('8.0000000000000000'), Decimal('30.26')),
 ('Hungary', 11, Decimal('7.1666666666666667'), Decimal('25.89')),
 ('Denmark', 7, Decimal('7.7142857142857143'), Decimal('14.63')),
 ('Belgium', 7, Decimal('9.0000000000000000'), Decimal('13.13')),
 ('USA', 6, Decimal('9.0000000000000000'), Decimal('14.54')),
 ('France', 21, Decimal('7.5384615384615385'), Decimal('46.59'))]